fpmsyncd/routesync.cpp

```diff
@@ -782,7 +782,7 @@ void RouteSync::onRouteMsg(int nlmsg_type, struct nl_object *obj, char *vrf)
             // In this case since we do not want the route with next hop on eth0/docker0, we return. 
             // But still we need to clear the route from the APPL_DB. Otherwise the APPL_DB and data 
             // path will be left with stale route entry
-            if(alsv.size() == 1)
+            if((nlmsg_type == RTM_DELROUTE) && (alsv.size() == 1))
```

Comment:
@prsunny, before PR #2553, when delete route update comes with only next hop on eth0/docker0, the update will not be sent to APPL_DB and ASIC_DB. This resulted in out-of-sync route table situation between kernel and ASIC_DB. The idea of PR #2553 is to avoid leaving these routes un-deleted in APPL_DB and ASIC_DB when the "delete" route update comes with all the next hops removed except next hop on eth0/docker0. For "add" route updates, due to timing/sequence of next hop updates, we may get a valid route update with a next hop on interfaces other than eth0/docker0 which will be sent to APPL_DB and ASIC_DB. Later if we receive add route update with next hop on eth0/docker0 (only one next hop), the changes in PR #2553 wrongly delets the already programmed valid route - as commented by @peter-yu. This is fixed by doing the delete only for delete route updates.